rfd/0192-oci-join.md

```diff
@@ -0,0 +1,282 @@
+  oracle:
+    allow:
+      - tenancy: "ocid1.tenancy.oc1..<unique ID>"  # the OCID for Alice's tenancy
+        parent_compartments: "ocid1.compartment.oc1..<unique ID>" # the OCID for Alice's compartment
+        # If needed, Alice can further restrict the compartments and regions
+        # instances can join from.
+```
+
+She would then create the provision token:
+
+```sh
+$ tctl create token.yaml
+```
+
+Next, Alice would install, configure, and start Teleport on all of her instances.
+If Alice has not yet created her instances, she can set `user_data` in each
+instance's metadata to add an init script for
+(cloud-init)[https://docs.oracle.com/en-us/iaas/Content/Compute/Tasks/launchinginstance.htm#].
+Otherwise, she can run the following script locally to install Teleport on her
+existing instances:
+
+```sh
+$ INSTANCE_IDS=$(oci compute instance list | jq -r '.data | map(.id) | join(" ")') # filter instances as needed
+$ for INSTANCE_ID in $(echo $INSTANCE_IDS)
+do
+  oci instance-agent command create \
+  --compartment-id <compartment-id> \
+  --content '{"source": {"sourceType": "TEXT", "text": "curl https://cdn.teleport.dev/install.sh | bash -s <Teleport version> && \
+  teleport node configure --token oci-token --join-method oracle --proxy example.com && \
+  sudo systemctl start teleport"}}' \
+  --target '{"instanceId": "$INSTANCE_ID"}'
+done
+```
+
+She can confirm that the nodes have joined either in the web UI or by running `tsh ls`.
+
+### Implementation
+
+#### Token spec
+
+The provision token will be extended to include a new `oracle` section:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: example-oci-token
+spec:
+  roles: [Node, Kube, Db]
+  oracle:
+    allow:
+        # OCID of the tenancy to allow instances to join from. Required.
+      - tenancy: "ocid1.tenancy.oc1..<unique ID>"
+        # OCIDs of compartments to allow instances to join from. Only the direct parent
+        # compartment applies; i.e. nested compartments are not taken into account.
+        # If compartments is empty,
+        # instances can join from any compartment in the tenancy.
+        parent_compartments: ["ocid1.compartment.oc1...<unique_ID>"]
+        # Regions to allow instances to join from. Both full names ("us-phoenix-1")
+        # and abbreviations ("phx") are allowed. If regions is empty, instances can join from any region.
+        regions: ["phx", "us-ashburn-1"]
+        # Add more entries as necessary.
+      - tenancy: "..."
+        parent_compartments: ["foo", "bar"]
+        regions: ["baz", "quux"]
+        # ...
+```
+
+#### Permissions
+
+Before the join process can begin, the nodes joining need permission to authenticate clients.
+
+- Create a [dynamic group](https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/managingdynamicgroups.htm)
+that matches all the instances that will join Teleport.
+- Create the following policy: `Allow dynamic-group <dynamic-group-name> to inspect authentication in tenancy`
```

Comment:
Have you been able to find much about what exactly this policy allows and if it seems acceptable to allow this for all nodes that need to join teleport? If it only allows you call the authenticateClient endpoint it seems okay to me, all it allows is for the node to get information about itself

Reply:
I haven't found anything that explicitly states what it allows, but these docs on the `authentication_oci` plugin for Oracle Heatwave ([1](https://docs.oracle.com/en-us/iaas/mysql-database/doc/mandatory-policies-and-permissions.html#MYAAS-GUID-A93275AA-3E3D-4172-BB9E-300406B14904), [2](https://docs.oracle.com/en-us/iaas/mysql-database/doc/prerequisites1.html)) suggest to me that it's only authenticateClient.